qsv: enable onevpl instead of libmfx

galinart opened the pull request:
Intel oneVPL is a successor to MediaSDK, but removed some obsolete
features of MediaSDK[1] including OPAQUE memory, multi-frame encode, user plugins and LA_EXT rate control mode
etc.

No regressions are expected and observed for existing HandBrake codebase, since oneVPL dispatcher expected to replace mfx_dispatcher smoothly
and have backward compatibility support[3].

All future Intel QSV features will utilize oneVPL interfaces[2].

More details:
[1] https://spec.oneapi.com/versions/latest/elements/oneVPL/source/appendix/VPL_intel_media_sdk.html
[2] https://github.com/oneapi-src/oneVPL
[3] https://software.intel.com/content/www/us/en/develop/articles/upgrading-from-msdk-to-onevpl.html

UPD: public FFMPEG oneVPL patch is under review https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=4885

galinart committed:
qsv: add onevpl patches into ffmpeg contrib

galinart committed:
qsv: add onevpl to contribs

oneVPL disaptcher[1] is successor of the MediaSDK dispatcher.
Third party mfx_dispatch[2] is not needed anymore.

[1] https://github.com/oneapi-src/oneVPL
[2] https://github.com/lu-zero/mfx_dispatch

galinart committed:
qsv: disable onevpl deprecated features for onevpl transition

galinart commented:
Hi @sr55,

Some help from your side would be much appreciated to add oneVPL tar.gz (https://github.com/oneapi-src/oneVPL/releases) into contrib repo. Currently it uses direct link to original repo.

Linux GitHub CI does not find oneVPL inlcudes, I believe some CI/Docker changes required as well to add it as dependency.
Let me know if changes required from my side.

sr55 commented:
Will take a look on both counts in a bit. Thanks :)

maximd33 commented:
it has been tested already intensively,
so the nightly might be great next step

sr55 commented:
Thanks max.

File uploaded:  https://github.com/HandBrake/HandBrake-contribs/releases/download/contribs/oneVPL-2021.6.0.tar.gz  

Checking on Linux build issue.  EDIT: Reproduced locally so it looks like a code issue. Trying to track it down.  libvpl has built ok it seems.

sr55 commented:
Looking at the structure of the build/contrib/include/vpl folder, it looks like the headers hsould be   #include vpl/mfxstructures.h

Looks like it impacts a few other files too. This would seem to match how it was before the rename.

Not really sure why it worked on Windows though.

galinart commented:
@sr55 the reason why is was changed to  #include "mfxstructures.h" (not  #include vpl/mfxstructures.h) from #include "mfx/mfxstructures.h" because of support 2 compilations options: against old mfx_disaptch and against onevpl. 
FFMPEG also supports both options for compilation.

I added the path for vpl folder since previously we had #include mfx/mfxstructures.h, but it seems that it works only for Windows:
https://github.com/HandBrake/HandBrake/pull/3870/commits/800e258df7c56c9fef5f148b332ef10d5bab0a61#diff-de8c1c4762a799d52ca0baf43ae1a4315bc5cf1be2f6fa6fd67fd0e4584ad9c2R125

maximd33 commented:
it should be minor change for switching to vpl/*,
agree

sr55 commented:
For us, we always ship ffmpeg and don't allow users to use their own, so we don't have a requirement to maintain backwards compatibility  if that makes anything easier. 

Trying to understand why what you did didn't work. The commands run indicate the path isn't being added on linux for some reason.

sr55 commented:
Found the problem.  libhb/modules.defs

The "LIBHB.GCC.I += $(CONTRIB.build/)include/vpl" line is within an ifeq cross compile block.  The indentation makes it super unclear. 

![image](https://user-images.githubusercontent.com/628593/133906175-1d6e065a-3ca7-4985-9a9f-bc3f4d54edfc.png)



Same problem for gtk gui I suspect or maybe just missing the include path for it in it's own modules.defs

galinart commented:
For simplicity, I will change #include mfx/mfxstructures.h to #include vpl/mfxstructures.h and it should work if maintaining backwards compatibility is not a requirement.

Mister-XY commented:
Little question, with this PR a file named mfxlib.log is created under c:\Temp\. 
Which is not really a problem, but unfortunately this file grows extremely. 
100 GB and larger is no problem.
Does the file really have to be?

galinart committed:
qsv: enable onevpl instead of libmfx

Intel oneVPL is a successor to MediaSDK, but removed some obsolete
features of MediaSDK[1] including OPAQUE memory, multi-frame encode, user plugins and LA_EXT rate control mode
etc.

No regressions are expected and observed for existing HandBrake codebase, since oneVPL dispatcher expected to replace mfx_dispatcher smoothly
and have backward compatibility support[3].

All future Intel QSV features will utilize oneVPL interfaces[2].

More details:
[1] https://spec.oneapi.com/versions/latest/elements/oneVPL/source/appendix/VPL_intel_media_sdk.html
[2] https://github.com/oneapi-src/oneVPL
[3] https://software.intel.com/content/www/us/en/develop/articles/upgrading-from-msdk-to-onevpl.html

maximd33 commented:
@Mister-XY, VPL path should not produce any log file like that
what is inside and is this clean & fresh build ?

galinart committed:
qsv: add cmake 3.16.3 as minimum supported version requirement for oneVPL

Mister-XY commented:
i have cmake 3.18.xx
I have uploadet a small log.
[mfxlib.log](https://github.com/HandBrake/HandBrake/files/7196816/mfxlib.log)

galinart commented:
@Mister-XY what driver version do you use?

Mister-XY commented:
30.0.100.9894 is the newest driver.

maximd33 commented:
@Mister-XY how exactly do you run HB?
I would like to reproduce the issue,
also, please try non-Beta driver , from https://www.intel.com/content/www/us/en/download/19344/intel-graphics-windows-10-windows-11-dch-drivers.html

in general - it is side effect, not really related to PR but thanks for details shared

Mister-XY commented:
I run via WinGui.
I have tested for you, is not a PR problem, is a driver problem.
with 30.0.100.9805 there is no log file and with the latest driver 30.0.100.9864, you have this log file.

galinart commented:
@Mister-XY thanks for the details. It is driver issue not exactly related to this PR, should be fixed soon.

sr55 commented:
Sorry for the delay. Managed to get some time in this morning and it's looking pretty good.

| Test Case | Notes |
|------|------|
| H.264 |  OK  ​|
| H.265 |  OK  ​|
| H.265 10-bit |  OK  ​|
| H.265 low power |  OK  ​|
| H.265 low power 10bit |  OK  ​|
| H.264 downscaling |  OK  ​|
| H.265 10bit, 10bit source |  OK  ​|
| Full Path Encoding | OK |

Looks like this is good functionality. I'll have a quick scan over the code and merge if there are no objections?

sr55 commented on contrib/libvpl/module.defs:
```diff
@@ -0,0 +1,50 @@
+$(eval $(call import.MODULE.defs,LIBVPL,libvpl))
+$(eval $(call import.CONTRIB.defs,LIBVPL))
+
+LIBVPL.FETCH.url     = https://github.com/oneapi-src/oneVPL/archive/refs/tags/v2021.6.0.tar.gz
```
You can add the HandBrake secondary URL here now. It should be uploaded to our HandBrake-contribs repo

galinart replied:
done

maximd33 commented:
>merge if there are no objections?
@sr55 , agree and thank you in advance

galinart committed:
qsv: add onevpl from contribs

galinart commented:
@Mister-XY new driver .9955 should solve log file problem https://www.intel.com/content/www/us/en/download/19344/intel-graphics-windows-dch-drivers.html

Mister-XY commented:
Thx for the info, i use the driver since a few days and have no problem. But i see, that Handbrake use the old api 1.35. That shows the activity log.

maximd33 commented:
@Mister-XY over time and with new functionality etc - API will be increased,
should be fine by now.
